### PR TITLE
Run armory without docker

### LINKED ISSUE #457: running armory without docker
Hi, some of use cannot install docker in our grids. Can you provide some instructions of how to manually configure our environment to be able to run armory without docker? e.g., list of python packages, environment variables, other dependencies, etc.
Thanks

### PR DESCRIPTION
**Pull request by ng390**
Closes #457

### CODE REVIEW

**Comment by ng390** [comment]
Tested on Ubuntu 18.04 LTS with python 3.6 on cifar10_baseline and librispeech_baseline_sincnet

File: armory/eval/evaluator.py [hunk 1]
```diff
@@ -144,8 +171,17 @@ def _delete_tmp(self):
...
-        container_port = 8888
         self._write_tmp()
+        if self.no_docker:
+            if jupyter or interactive or command:
+                raise ValueError(
+                    "jupyter, interactive, or bash commands only supported when running Docker containers."
+                )
+            runner = self.manager.start_armory_instance(envs=self.extra_env_vars)
+            self._run_config(runner)
+            return
```
Comment (davidslater) [inline armory/eval/evaluator.py:182]: You need to do the `self._delete_tmp()` cleanup before returning.

Comment (davidslater) [reply armory/eval/evaluator.py:182]: Similar logging to below would also be helpful.

**Comment by davidslater** [comment]
Is there some sort of automated test that we can set up for this? Seems difficult to tell if/when we introduce a regression bug down the line.

**Comment by davidslater** [comment]
In looking through this and considering the AWS evaluations (as well as the fact that evaluations are now done as scenarios and not explicitly in the Evaluator class), maybe we want to consider restructuring how the `Evaluator` and `Management` classes work and interact with each other (and how `__main__` interacts with them). @seanpmorgan thoughts?

**Comment by davidslater** [comment]
Here's a thought: could we run armory recursively in "host" mode inside of an armory docker container to test?
```
python -m armory exec tf1 -- armory run --no-docker ...
```

**Comment by seanpmorgan** [comment]
> Here's a thought: could we run armory recursively in "host" mode inside of an armory docker container to test?
> 
> ```
> python -m armory exec tf1 -- armory run --no-docker ...
> ```

Not quite following why that's needed. Can't we just run this on a fresh ubuntu VM in the CI?

**Comment by davidslater** [comment]
> > Here's a thought: could we run armory recursively in "host" mode inside of an armory docker container to test?
> > ```
> > python -m armory exec tf1 -- armory run --no-docker ...
> > ```
> 
> Not quite following why that's needed. Can't we just run this on a fresh ubuntu VM in the CI?

Wouldn't we then have to work out the full build process for python, cuda, tensorflow, etc., on the fresh ubuntu VM? Aren't we trying to avoid supporting that?

**Comment by ng390** [comment]
> > > Here's a thought: could we run armory recursively in "host" mode inside of an armory docker container to test?
> > > ```
> > > python -m armory exec tf1 -- armory run --no-docker ...
> > > ```
> > 
> > 
> > Not quite following why that's needed. Can't we just run this on a fresh ubuntu VM in the CI?
> 
> Wouldn't we then have to work out the full build process for python, cuda, tensorflow, etc., on the fresh ubuntu VM? Aren't we trying to avoid supporting that?

armory exec will run into permissions issues trying to access directories on the container in host mode, e.g. trying to read the default config in .armory. This can be circumvented by adding a flag to not set container_args["user"] or to docker run the command, e.g.
```
docker run -w /armory_dev twosixarmory/tf1:${version} python -m armory run --no-docker tests/scenarios/tf1/image_classification.json
```

**Comment by AngusG** [comment]
Hi, thanks for this PR. I'm on an academic cluster with docker but no root access and was having permissions issues so this is helpful. 

I confirmed this PR works on CIFAR-10 and librispeech, but wondering how to specify the pytorch/tensorflow version which was previously supplied to `docker_image": "twosixarmory/pytorch:0.6.0",`. I'm assuming is ignored when using `--no-docker`. The cifar10 example seems to use aspects of pytorch and tensorflow at the same time, but defaults to tensorflow for evaluation. For example: 

```
armory run --no-docker official_scenario_configs/cifar10_baseline.json
...
2020-04-25 22:15:43 ws14 art.classifiers.pytorch[17952] INFO Inferred 1 hidden layers on PyTorch classifier.
...
tensorflow/core/platform/cpu_feature_guard.cc:142] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 FMA
```
Is it possible to have it use one or the other, and is there a flag to do so? Thanks!

**Comment by davidslater** [comment]
> Hi, thanks for this PR. I'm on an academic cluster with docker but no root access and was having permissions issues so this is helpful.
> 
> I confirmed this PR works on CIFAR-10 and librispeech, but wondering how to specify the pytorch/tensorflow version which was previously supplied to `docker_image": "twosixarmory/pytorch:0.6.0",`. I'm assuming is ignored when using `--no-docker`. The cifar10 example seems to use aspects of pytorch and tensorflow at the same time, but defaults to tensorflow for evaluation. For example:
> 
> ```
> armory run --no-docker official_scenario_configs/cifar10_baseline.json
> ...
> 2020-04-25 22:15:43 ws14 art.classifiers.pytorch[17952] INFO Inferred 1 hidden layers on PyTorch classifier.
> ...
> tensorflow/core/platform/cpu_feature_guard.cc:142] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 FMA
> ```
> 
> Is it possible to have it use one or the other, and is there a flag to do so? Thanks!

In the config, you should be able to set `"docker_image": null`.

In terms of PyTorch and TensorFlow, we use TFDS under the hood for data loading, which is why TensorFlow is showing up. We are working on a native PyTorch data loader, but there is still some work to be done on that front. For evaluation, the framework used is based on the model selected.

**Pull request merged**